### PR TITLE
Autonomous gateway

### LINKED ISSUE #636: Autonomous WebGateway
This story is about the creation of the next version of the TFGateway (TFGW).

## New Features:
The TFGW become autonomous the same way 0-OS node are. Instead of exposing an API the TFGW will be driven by reservation. Some new primitives will need to be created to be able to configure the web gateway.

The new primitives given by the TFGW:
- TCP proxy:
  - proxy TLS TCP traffic to a specific IP
  - reverse TCP proxy using [tcp router client](https://github.com/threefoldtech/tcprouter#reverse-tunneling)
- DNS
  - get a subdomain of the domain managed by the TFGW
  - Allow the user to delegate part of his own domain to the TFGW

## TFGW identity
Update explorer so it support a new entity in the grid, the webgateway

A gateway can use the same logic as a node to generate its identity. 
Generate a random seed, generate a keypair using ed25119 algo on the seed, then base58 encode the public key. This is the ID of the gateway.
see https://github.com/threefoldtech/zos/tree/master/pkg/identity

The TFGW type will only provide Network Units (?)

TFGW needs to report:
- uptime
- location
- managed domains
- port to use for the tcp router client
- name of the DNS: A user needs to know how he needs to configure its DNS record to let TFGW serve subdomain of the user domain
- nameserver IPs
- amount of primitives deployed for each type

### TCP proxy to an IP

```toml
domain = (S)
destination = (S) # format 'ip:port'
```

### Reserve proxy using TCP server client

```toml
domain = (S)
secret = (S)
```

notes: the format of the secret MUST be `threebotID-secret`. If the secret is not prefixed with the customer threebot-id. we reject the reservation. This allow the secret to not overlap between clients.
This is done like this so the secret is known before the reservation is send. So you can define a container and it's GW config in the same reservation.

### Subdomain of a domain (TFGW managed domain or user delegated domain)

```
domain = (S) # full domain. e.g: `user1.tfgw1.io` if `tfgw1.io` is the managed domain of the TFGW
ip = (LS) # list of target IP
```

### delegate a domain to the TFGW

```
domain = (S)
```

### Provisioning routine

This routine is responsible to watch for new reservation. This is done by polling the explorer.

The reservation is validated, then converted into the proper configuration for both TCP router and CoreDNS.

This component is also responsible to decommission workloads once they expired

### Identity routine

Responsible to register the TFGW into the explorer and keep sending heartbeat and update information (amount of workloads deployed, used network units...)

## Questions

- What units are going to be used to reserve TFGW
- Should we try to bundle TCP Router, CoreDNS and the rest of the code into one single binary ? Make it super easy to deploy

## Tasks
- [x] Update explorer so it support a new entity in the grid, the webgateway
- [x] Define reservation schema for all the different primitive provided by the web gateway
- [x] Implement TFGW

### PR DESCRIPTION
**Pull request by zaibon**
fixes https://github.com/threefoldtech/home/issues/636

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+	rs := make([]*provision.Reservation, 0, len(ids))
+	for _, id := range ids {
+		r, err := s.get(string(id))
+		if err != nil {
+			return nil, err
+		}
+
+		if r.Expired() {
+			// r.Tag = Tag{"source": "FSStore"}
```
Comment (LeeSmet) [inline cache/cache.go]: forgot to remove?

File: cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	var reservation provision.Reservation
+
+	if validV1(reader.Version()) {
+		if err := json.NewDecoder(reader).Decode(&reservation); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("unknown reservation object version (%s)", reader.Version())
+	}
+	// reservation.Tag = Tag{"source": "FSStore"}
```
Comment (LeeSmet) [inline cache/cache.go]: forgot to remove?

File: cache/cache_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		wantErr bool
+	}{
+		{
+			name: "main",
+			args: args{
+				r: &provision.Reservation{
+					ID:       "r-1",
+					Created:  time.Now().UTC().Add(-time.Minute).Round(time.Second),
+					Duration: time.Second * 10,
+					// Tag:      Tag{"source": "FSStore"},
```
Comment (LeeSmet) [inline cache/cache_test.go]: forgot to remove?

File: cache/cache.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+)
+
+var (
+	// reservationSchemaV1 reservation schema version 1
+	reservationSchemaV1 = versioned.MustParse("1.0.0")
+	// ReservationSchemaLastVersion link to latest version
+	reservationSchemaLastVersion = reservationSchemaV1
+)
+
+const _reservationsKey = "tfgateway_reservations"
```
Comment (LeeSmet) [inline cache/cache.go]: is there a particular reason to this variable starts with an underscore? :P

File: cmd/tfgateway/main.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+		PublicKeyHex: hex.EncodeToString(kp.PublicKey),
+		OsVersion:    version.Current().Short(),
+		Location: directory.Location{
+			Continent: loc.Continent,
+			Country:   loc.Country,
+			City:      loc.City,
+			Longitude: loc.Longitute,
+			Latitude:  loc.Latitude,
+		},
+		// ManagedDomains: ,
```
Comment (LeeSmet) [inline cmd/tfgateway/main.go]: forgot to remove?

File: counters.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		// u, err = processKubernetes(r)
+	}
+	// if err != nil {
+	// 	return err
+	// }
+
+	// c.CRU.Increment(u.CRU)
+	// c.MRU.Increment(u.MRU)
+	// c.SRU.Increment(u.SRU)
+	// c.HRU.Increment(u.HRU)
```
Comment (LeeSmet) [inline counters.go]: commented code :)

File: counters.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+	return nil
+}
+
+// type resourceUnits struct {
+// 	SRU uint64 `json:"sru,omitempty"`
+// 	HRU uint64 `json:"hru,omitempty"`
+// 	MRU uint64 `json:"mru,omitempty"`
+// 	CRU uint64 `json:"cru,omitempty"`
+// }
```
Comment (LeeSmet) [inline counters.go]: same here

File: go.mod [hunk 8]
```diff
@@ -0,0 +1,27 @@
+module github.com/threefoldtech/tfgateway
+
+go 1.14
+
+replace github.com/threefoldtech/zos => ../zos
+
+replace github.com/threefoldtech/tfexplorer => ../tfexplorer
```
Comment (LeeSmet) [inline go.mod]: Are we sure we want local dependencies?

File: reservation_cache.go [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+	var reservation provision.Reservation
+
+	if validV1(reader.Version()) {
+		if err := json.NewDecoder(reader).Decode(&reservation); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("unknown reservation object version (%s)", reader.Version())
+	}
+	// reservation.Tag = Tag{"source": "FSStore"}
```
Comment (LeeSmet) [inline reservation_cache.go:208]: commented code

File: reservation_cache.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+				return nil, os.Remove(path.Join(s.root, info.Name()))
+			}
+		}
+
+		r, err := s.get(info.Name())
+		if err != nil {
+			return nil, err
+		}
+		if r.Expired() {
+			// r.Tag = Tag{"source": "FSStore"}
```
Comment (LeeSmet) [inline reservation_cache.go:146]: same here

File: provision.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+func WorkloadToProvisionType(w workloads.ReservationWorkload) (*provision.Reservation, error) {
+
+	reservation := &provision.Reservation{
+		ID:        w.WorkloadId,
+		User:      w.User,
+		Type:      provision.ReservationType(w.Type.String()),
+		Created:   w.Created.Time,
+		Duration:  time.Duration(w.Duration) * time.Second,
+		Signature: []byte(w.Signature),
+		// Data:      w.Content,
```
Comment (LeeSmet) [inline provision.go]: same here

**Review by LeeSmet — APPROVED** [review]
LGTM, quite some commented code which should probably cleaned up :)

File: go.mod (hunk 8 above)
Comment (zaibon) [reply go.mod]: we don't but I need a version released on tfexplore first.

**Pull request merged**